### DISCUSSION TITLE
Testing elementTools

### DISCUSSION THREAD

**Discussion by malomiquel** [clientIO/joint#1897] [Q&A]
Hello, 

I want to test methods inside elementTools but I don't know if is it possible.
In the Control: test getPosition and setPosition  functions
In the Remove Button: test the action method 
I work with ReactJS and jest for the code

Here is my code:

![code](https://user-images.githubusercontent.com/83480447/202512424-922606a7-7f92-4860-a4c6-31bc18777136.png)

Thank you

**Comment by jamesgeorgewilliams** [comment]
Hi,

It is certainly possible to test these, but depending on what you want to do, it might be easier to introduce some new technology to your testing.

`Jest` uses `JSDOM` which doesn't have a layout engine. That means if you want to test things like `position` or `size` in your application, it will be difficult to do so without mocking functionality in Jest.

Introducing something like `puppeteer` or `playwright` to work with Jest would probably make your life easier. 

Using `puppeteer` and `Jest`, you could test `elementTools` like the following:

```js
const ResizeToolLeft = elementTools.Control.extend({
    getPosition: function() {
      return { x: 0, y: 0 };
    },
    setPosition: function(view, coordinates) {
      const model = view.model;
      model.resize(
        Math.max(60, model.size().width - coordinates.x),
        model.size().height,
        {
          direction: "left"
        }
      );
    }
  });
```

```js    
it('renders elementTools', () => {
   expect(page.waitForSelector('.joint-tool')).toBeTruthy;
});

it('elementTools Resize', async () => {

    await page.waitForSelector('.joint-type-standard-rectangle');

    const rect = await page.$('.joint-type-standard-rectangle');
    const rectBBox = await rect.boundingBox();

    const resizeTool = await page.$('.joint-tool');
    const resizeToolBBox = await resizeTool.boundingBox();

    /* 
       getPosition - Test Element Tool is in correct position 
    */

    // 6 - default radius of tool circle
    expect(rectBBox.x).toBe(resizeToolBBox.x + 6);
    expect(rectBBox.y).toBe(resizeToolBBox.y + 6);

    /* 
       setPosition - Resize element on x axis to the left by its current width
    */

    // Mousedown on center of Resize tool
    await page.mouse.move(resizeToolBBox.x + resizeToolBBox.width / 2, resizeToolBBox.y + resizeToolBBox.height / 2);
    await page.mouse.down();
    await page.mouse.move(rectBBox.x - rectBBox.width, 0);
    await page.mouse.up();

    // Get new Rectangle position
    const rectBBox2 = await rect.boundingBox();

    // Expect original X position to be greater
    expect(rectBBox.x).toBeGreaterThan(rectBBox2.x);
    // Expect original width to be smaller
    expect(rectBBox.width).toBeLessThan(rectBBox2.width);

     /* 
       setPosition - Resize element on x axis to the right by twice its current width
     */

    // Get new tools position
    const resizeToolBBox1 = await resizeTool.boundingBox();

    await page.mouse.move(resizeToolBBox1.x + resizeToolBBox1.width / 2, resizeToolBBox1.y + resizeToolBBox1.height / 2);
    await page.mouse.down();
    await page.mouse.move(rectBBox2.x + rectBBox2.width * 2, 0);
    await page.mouse.up();

    // Get new Rectangle position
    const rectBBox3 = await rect.boundingBox();

    // Expect new X position to be greater than the original
    expect(rectBBox3.x).toBeGreaterThan(rectBBox.x);
    // Expect new Rectangle to maintain width above our minimum resize value
    expect(rectBBox3.width).toBeGreaterThanOrEqual(60);
});
```

**Reply by malomiquel** [reply]
Thanks for your answer but is it possible to mock elementTools only with jest ?

**Reply by kumilingus** [reply]
You can call the tool's `setPosition()` directly and then check the new size and position of the model (not the view).

**Reply by malomiquel** [reply]
Hello,

When I try to mock elementTools I have this error:

<img width="345" alt="Capture d’écran 2022-11-23 100219" src="https://user-images.githubusercontent.com/83480447/203507056-759dc21b-e844-417e-bd75-8e2eab291d69.png">

Here is my code to mock elementTools: 

![code](https://user-images.githubusercontent.com/83480447/203507691-69f5c593-4a25-4b4b-8420-305b8123d41c.png)

Thank you

**Reply by kumilingus** [reply]
The `ResizeToolLeft` class was defined by you. You need to import it into the test (like any other function you're planning to test).
